Feature: use last matching mapping for linking

piptouque opened the pull request:
Hi!
Thanks for dfm, I've been using to set-up a few GNU/Linux machines and I'm liking it so far. 

### Rationale

So I've been wanting to include some third-party repos into my profile
(mainly plug-in managers such as [vim-plug](https://github.com/junegunn/vim-plug)) and I've been adding them as modules.
Since installing them amounts to simply copying a few files,
I found it useful to only link _some_ files in, and skip the rest.
Using this change, one can do the above like so:

```yml
# .dfm.yml for my profile
modules:
  - repo: my_repo
  mappings:
    # my catch-all pattern
    - match: '.*'
      skip: true
   # only my_file will be linked
    - match: my_file
       target_dit: ~/.my_config_dir
```


I'm not convinced by the picking of the last one instead of the first,
but it matches previous behaviour more closely when there was no `skip` or `link_as_dir` option in the module.

###  Before

- `generate_link` would to skip all mappings`
if a single `skip`option was found for the module,
- it would return immediately if `link_as_dir` was found,
- otherwise, it would pick the last matching mapping.

### Now

- `generate_link` always picks the last of matching mappings.

Commit:
feature: use last matching mapping for linking

Before:
- `generate_link` would to skip all mappings`
if a single `should_skip`option was found,
- it would return immediately if `link_as_dir` was found,
- otherwise, it would pick the last matching mapping.

Now `generate_link` always picks the last of matching mappings.

chasinglogic commented:
Hi @piptouque thanks for the contribution!

I'd like to see some new tests for this behaviour.

The code is a bit obtuse and I have some potential style improvements but first I'd like to see tests so we can make sure my suggestions don't break your intended new behaviour.

Commit:
[fix]: config paths and variable on windows.

Commit:
[fix]: paths in replace()

Commit:
[+]: Use GitPython for internal git usage.

Profile delegates all git-related stuff to DotfileRepository now.
Logging with progress bars and diffs must be reimplemented.
CLI commands were not changed.

chasinglogic commented:
@piptouque this PR is quickly spiraling into a rewrite. I have no intention of merging the use of `GitPython`. I intentionally didn't use it because keeping a low dependency count is important to me as a Python application distributor it means less things break and mess with user's system python.

Commit:
[fix]: handling of repo when not set.

Commit:
[fix]: no is_relative_to for backward compatibility

piptouque commented:
Hi there, sorry for the late reply, I hadn't realised my more recent changes would be reflected in this PR. I understand that you want to keep it simple. Thanks for the feedback, have a nice day!
